File: frontend/src/features/projects/components/cad/survey/terrain.ts
/**
 * Pure-TypeScript terrain analysis over a TIN (slope, aspect, 3D surface area
 * and whole-surface statistics).
 *
 * Mirrors the Rust `survey-core::terrain` module exactly and serves as the
 * guaranteed fallback when the WebAssembly module is unavailable. Both paths
 * are numerically equivalent.
 *
 * Conventions match `surface.ts`: coordinates are Y/Northing (`n`) /
 * X/Easting (`e`) / Z/Elevation (`z`); all angles are in degrees, and aspect
 * is an azimuth measured clockwise from North (0 = faces North, 90 = East).
 */

import type { SurfaceTin } from "./surface.ts";

const DEG = 180 / Math.PI;

/** Slope / aspect / area facts for a single TIN triangle. */
export interface TriangleAnalysis {
  /** Index of the triangle within `SurfaceTin.triangles`. */
  index: number;
  /** Slope angle from horizontal, degrees (0 = flat, 90 = vertical). */
  slopeDeg: number;
  /** Slope as a percentage grade (rise/run × 100). */
  slopePercent: number;
  /**
   * Aspect: downslope-facing azimuth in degrees clockwise from North
   * (0 = faces North, 90 = East). `null` for a flat triangle.
   */
  aspectDeg: number | null;
  /** Plan (projected) area, m². */
  planArea: number;
  /** True 3D surface area, m² (>= plan area). */
  surfaceArea: number;
}

/** Whole-surface terrain statistics aggregated from the per-triangle analysis. */
export interface TerrainStats {
  planArea: number;
  surfaceArea: number;
  meanSlopeDeg: number;
  minSlopeDeg: number;
  maxSlopeDeg: number;
  minElevation: number;
  maxElevation: number;
  triangles: number;
}

function triangleArea2d(
  ax: number,
  ay: number,
  bx: number,
  by: number,
  cx: number,
  cy: number,
): number {
  return Math.abs((bx - ax) * (cy - ay) - (cx - ax) * (by - ay)) * 0.5;
}

/** Analyse every triangle of the TIN for slope, aspect and area. */
export function analyseTriangles(tin: SurfaceTin): TriangleAnalysis[] {
  const out: TriangleAnalysis[] = [];
  tin.triangles.forEach((t, index) => {
    const a = tin.points[t.a];
    const b = tin.points[t.b];
    const c = tin.points[t.c];
    if (!a || !b || !c) return;

    // Edge vectors in (E, N, Z).
    const u = [b.e - a.e, b.n - a.n, b.z - a.z] as const;
    const v = [c.e - a.e, c.n - a.n, c.z - a.z] as const;

    // Normal = u × v.
    const nx = u[1] * v[2] - u[2] * v[1];
    const ny = u[2] * v[0] - u[0] * v[2];
    const nz = u[0] * v[1] - u[1] * v[0];
    const nlen = Math.sqrt(nx * nx + ny * ny + nz * nz);

    const planArea = triangleArea2d(a.e, a.n, b.e, b.n, c.e, c.n);
    const surfaceArea = nlen / 2;

    let slopeDeg = 0;
    let aspectDeg: number | null = null;
    if (nlen >= 1e-12) {
      const cosSlope = Math.abs(nz) / nlen;
      slopeDeg = Math.acos(Math.min(1, Math.max(-1, cosSlope))) * DEG;

      // Orient the normal up so its horizontal projection points downslope.
      const dx = nz >= 0 ? nx : -nx;
      const dy = nz >= 0 ? ny : -ny;
      if (Math.abs(dx) >= 1e-12 || Math.abs(dy) >= 1e-12) {
        let az = Math.atan2(dx, dy) * DEG;
        if (az < 0) az += 360;
        aspectDeg = az;
      }
    }

    out.push({
      index,
      slopeDeg,
      slopePercent: Math.tan(slopeDeg / DEG) * 100,
      aspectDeg,
      planArea,
      surfaceArea,
    });
  });
  return out;
}

/** Aggregate whole-surface statistics from a TIN, or null when empty. */
export function terrainStats(tin: SurfaceTin): TerrainStats | null {
  if (tin.triangles.length === 0 || tin.points.length === 0) return null;
  const tris = analyseTriangles(tin);

  let planArea = 0;
  let surfaceArea = 0;
  let weightedSlope = 0;
  let minSlope = Infinity;
  let maxSlope = -Infinity;
  for (const t of tris) {
    planArea += t.planArea;
    surfaceArea += t.surfaceArea;
    weightedSlope += t.slopeDeg * t.planArea;
    minSlope = Math.min(minSlope, t.slopeDeg);
    maxSlope = Math.max(maxSlope, t.slopeDeg);
  }

  let minElevation = Infinity;
  let maxElevation = -Infinity;
  for (const p of tin.points) {
    minElevation = Math.min(minElevation, p.z);
    maxElevation = Math.max(maxElevation, p.z);
  }

  const meanSlopeDeg = planArea > 0 ? weightedSlope / planArea : 0;

  return {
    planArea,
    surfaceArea,
    meanSlopeDeg,
    minSlopeDeg: Number.isFinite(minSlope) ? minSlope : 0,
    maxSlopeDeg: Number.isFinite(maxSlope) ? maxSlope : 0,
    minElevation,
    maxElevation,
    triangles: tris.length,
  };
}

/**
 * Map a slope angle (degrees) to a colour on a green→yellow→red ramp, for
 * slope-shaded terrain rendering. `maxSlope` sets the top of the scale.
 */
export function slopeColor(slopeDeg: number, maxSlope: number): string {
  const denom = maxSlope > 0 ? maxSlope : 1;
  const t = Math.min(1, Math.max(0, slopeDeg / denom));
  // 0 → green (120°), 0.5 → yellow (60°), 1 → red (0°).
  const hue = 120 * (1 - t);
  return `hsl(${hue.toFixed(0)}, 80%, 45%)`;
}

File: backend/crates/survey-core/src/terrain.rs
//! Terrain analysis over a TIN: per-triangle slope, aspect, 3D surface area and
//! whole-surface statistics.
//!
//! These are standard digital-terrain-model deliverables in engineering survey
//! (slope shading, drainage/aspect analysis, true surface area for earthworks
//! and landscaping quantities). All angles are in degrees; X = Easting,
//! Y = Northing, Z = elevation.

use crate::tin::triangle_area_2d;
use crate::Tin;

const DEG: f64 = 180.0 / std::f64::consts::PI;

/// Slope/aspect/area facts for a single TIN triangle.
#[derive(Debug, Clone, Copy, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct TriangleAnalysis {
    /// Index of the triangle within `Tin::triangles`.
    pub index: usize,
    /// Slope angle from horizontal, degrees (0 = flat, 90 = vertical).
    pub slope_deg: f64,
    /// Slope as a percentage grade (rise/run × 100).
    pub slope_percent: f64,
    /// Aspect: downslope-facing azimuth in degrees clockwise from North
    /// (0 = faces North, 90 = East). `None` for a flat triangle.
    pub aspect_deg: Option<f64>,
    /// Plan (projected) area, m².
    pub plan_area: f64,
    /// True 3D surface area, m² (>= plan area).
    pub surface_area: f64,
}

/// Whole-surface terrain statistics aggregated from the per-triangle analysis.
#[derive(Debug, Clone, Copy, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct TerrainStats {
    /// Total plan (projected) area, m².
    pub plan_area: f64,
    /// Total true 3D surface area, m².
    pub surface_area: f64,
    /// Area-weighted mean slope, degrees.
    pub mean_slope_deg: f64,
    /// Minimum triangle slope, degrees.
    pub min_slope_deg: f64,
    /// Maximum triangle slope, degrees.
    pub max_slope_deg: f64,
    /// Minimum elevation across all points.
    pub min_elevation: f64,
    /// Maximum elevation across all points.
    pub max_elevation: f64,
    /// Number of triangles analysed.
    pub triangles: usize,
}

/// Analyse every triangle of the TIN for slope, aspect and area.
pub fn analyse_triangles(tin: &Tin) -> Vec<TriangleAnalysis> {
    let mut out = Vec::with_capacity(tin.triangles.len());
    for (index, t) in tin.triangles.iter().enumerate() {
        let a = tin.points[t.a];
        let b = tin.points[t.b];
        let c = tin.points[t.c];

        // Edge vectors in (E, N, Z).
        let u = (b.e - a.e, b.n - a.n, b.z - a.z);
        let v = (c.e - a.e, c.n - a.n, c.z - a.z);

        // Normal = u × v.
        let nx = u.1 * v.2 - u.2 * v.1;
        let ny = u.2 * v.0 - u.0 * v.2;
        let nz = u.0 * v.1 - u.1 * v.0;
        let nlen = (nx * nx + ny * ny + nz * nz).sqrt();

        let plan_area = triangle_area_2d(a.e, a.n, b.e, b.n, c.e, c.n);
        // 3D surface area = half the magnitude of the cross product.
        let surface_area = nlen / 2.0;

        // Slope = angle between the face normal and vertical (Z) axis.
        let (slope_deg, aspect_deg) = if nlen < 1e-12 {
            (0.0, None)
        } else {
            let cos_slope = (nz.abs()) / nlen;
            let slope = cos_slope.clamp(-1.0, 1.0).acos() * DEG;

            // Aspect = azimuth of the downslope (steepest-descent) direction.
            // For an up-pointing face normal, its horizontal projection already
            // points downslope (toward lower ground), so orient the normal up
            // and use that horizontal component directly.
            let (dx, dy) = if nz >= 0.0 { (nx, ny) } else { (-nx, -ny) };
            if dx.abs() < 1e-12 && dy.abs() < 1e-12 {
                (slope, None)
            } else {
                // Azimuth clockwise from North: atan2(East, North).
                let mut az = dx.atan2(dy) * DEG;
                if az < 0.0 {
                    az += 360.0;
                }
                (slope, Some(az))
            }
        };

        out.push(TriangleAnalysis {
            index,
            slope_deg,
            slope_percent: slope_deg.to_radians().tan() * 100.0,
            aspect_deg,
            plan_area,
            surface_area,
        });
    }
    out
}

/// Aggregate whole-surface statistics from a TIN.
pub fn terrain_stats(tin: &Tin) -> Option<TerrainStats> {
    if tin.triangles.is_empty() || tin.points.is_empty() {
        return None;
    }
    let tris = analyse_triangles(tin);

    let mut plan_area = 0.0;
    let mut surface_area = 0.0;
    let mut weighted_slope = 0.0;
    let mut min_slope = f64::INFINITY;
    let mut max_slope = f64::NEG_INFINITY;

    for t in &tris {
        plan_area += t.plan_area;
        surface_area += t.surface_area;
        weighted_slope += t.slope_deg * t.plan_area;
        min_slope = min_slope.min(t.slope_deg);
        max_slope = max_slope.max(t.slope_deg);
    }

    let (mut min_elevation, mut max_elevation) = (f64::INFINITY, f64::NEG_INFINITY);
    for p in &tin.points {
        min_elevation = min_elevation.min(p.z);
        max_elevation = max_elevation.max(p.z);
    }

    let mean_slope_deg = if plan_area > 0.0 {
        weighted_slope / plan_area
    } else {
        0.0
    };

    Some(TerrainStats {
        plan_area,
        surface_area,
        mean_slope_deg,
        min_slope_deg: if min_slope.is_finite() { min_slope } else { 0.0 },
        max_slope_deg: if max_slope.is_finite() { max_slope } else { 0.0 },
        min_elevation,
        max_elevation,
        triangles: tris.len(),
    })
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tin::build_tin;
    use crate::Point3;

    /// Flat surface at Z=5: slope 0, surface area == plan area.
    fn flat() -> Tin {
        build_tin(&[
            Point3::new(0.0, 0.0, 5.0),
            Point3::new(0.0, 10.0, 5.0),
            Point3::new(10.0, 10.0, 5.0),
            Point3::new(10.0, 0.0, 5.0),
        ])
        .unwrap()
    }

    /// Plane tilted so Z rises 1:1 toward increasing Easting: a 45° slope whose
    /// downhill (steepest-descent) direction faces West (aspect 270°).
    /// Points are (N, E, Z); Z == E here.
    fn ramp_east() -> Tin {
        build_tin(&[
            Point3::new(0.0, 0.0, 0.0),
            Point3::new(10.0, 0.0, 0.0),
            Point3::new(0.0, 10.0, 10.0),
            Point3::new(10.0, 10.0, 10.0),
        ])
        .unwrap()
    }

    #[test]
    fn flat_surface_is_zero_slope() {
        let stats = terrain_stats(&flat()).unwrap();
        assert!(stats.max_slope_deg.abs() < 1e-9);
        assert!((stats.surface_area - stats.plan_area).abs() < 1e-6);
        assert!((stats.plan_area - 100.0).abs() < 1e-6);
    }

    #[test]
    fn ramp_is_45_degrees_facing_west() {
        let tris = analyse_triangles(&ramp_east());
        for t in &tris {
            assert!((t.slope_deg - 45.0).abs() < 1e-6, "slope {}", t.slope_deg);
            // Z increases toward +E, so downslope faces West -> aspect ~270°.
            let asp = t.aspect_deg.unwrap();
            assert!((asp - 270.0).abs() < 1e-6, "aspect {asp}");
        }
        // Surface area = plan area / cos(45°) = 100 / 0.7071 ≈ 141.42.
        let stats = terrain_stats(&ramp_east()).unwrap();
        assert!((stats.surface_area - 141.42135).abs() < 1e-3);
        assert!((stats.mean_slope_deg - 45.0).abs() < 1e-6);
    }

    #[test]
    fn empty_tin_has_no_stats() {
        assert!(build_tin(&[Point3::new(0.0, 0.0, 0.0)]).is_err());
    }
}
